File: gherkin/javascript/src/legacy/gherkin/ast_node.ts
import { RuleType, TokenType } from './parser'
import Token from './token'

export default class AstNode {
  private readonly subItems = new Map<any, any[]>()

  constructor(public readonly ruleType: RuleType) {}

  public add(type: any, obj: any) {
    let items = this.subItems.get(type)
    if (items === undefined) {
      items = []
      this.subItems.set(type, items)
    }
    items.push(obj)
  }

  public getSingle(ruleType: RuleType) {
    return (this.subItems.get(ruleType) || [])[0]
  }

  public getItems(ruleType: RuleType) {
    return this.subItems.get(ruleType) || []
  }

  public getToken(tokenType: TokenType) {
    return (this.subItems.get(tokenType) || [])[0]
  }

  public getTokens(tokenType: TokenType): Token[] {
    return this.subItems.get(tokenType) || []
  }
}

File: gherkin/dotnet/Gherkin/AstNode.cs
using System.Collections.Generic;
using System.Linq;

namespace Gherkin
{
    public class AstNode
    {
        private readonly Dictionary<RuleType, IList<object>> subItems = new Dictionary<RuleType, IList<object>>();
        public RuleType RuleType { get; private set; }

        public AstNode(RuleType ruleType)
        {
            this.RuleType = ruleType;
        }

        public Token GetToken(TokenType tokenType)
        {
            return GetSingle<Token>((RuleType)tokenType);
        }

        public IEnumerable<Token> GetTokens(TokenType tokenType)
        {
            return GetItems<Token>((RuleType)tokenType);
        }

        public T GetSingle<T>(RuleType ruleType)
        {
            return GetItems<T>(ruleType).SingleOrDefault();
        }

        public IEnumerable<T> GetItems<T>(RuleType ruleType)
        {
            IList<object> items;
            if (!subItems.TryGetValue(ruleType, out items))
            {
                return Enumerable.Empty<T>();
            }
            return items.Cast<T>();
        }

        public void SetSingle<T>(RuleType ruleType, T value)
        {
            subItems[ruleType] = new object[] { value };
        }

        public void AddRange<T>(RuleType ruleType, IEnumerable<T> values)
        {
            foreach (var value in values)
            {
                Add(ruleType, value);
            }
        }

        public void Add<T>(RuleType ruleType, T obj)
        {
            IList<object> items;
            if (!subItems.TryGetValue(ruleType, out items))
            {
                items = new List<object>();
                subItems.Add(ruleType, items);
            }
            items.Add(obj);
        }
    }
}
